Decouple server initialization when creating a new process

thodkatz opened the pull request:
I noticed that when we attempt to create a new process with `start_model_session_process`, the inputs required to initialize the server are sent to the process (serialized), and the construction of the server's dependencies (`PredictionPipeline`) is taking place on the process before launching the server.

One issue with this approach is that if the server initialization fails, we can't utilize the error handling of the server, and the process will die without any client feedback.

Thus, I moved any logic that is responsible to initialize the server to a separate `init` called, so when the client triggers the init process, we can be notified if something went wrong, and potentially recovering from it as well.

I have refactored the `Session` as well, using type vars for the type of the client, instead of the hard-coded `BioModelClient`.

thodkatz commented on tiktorch/server/session/process.py:
```diff
@@ -38,7 +39,7 @@ def _check_shape(self, tensor_id: str, tensor: Tensor):
             target_axis_size = tensor.sizes[axis.id]
             if axis.id not in tensor.sizes:
-                ValueError(f"{axis.id} not found in {tensor.sizes}")
+                ValueError(f"{axis.id} found not in {tensor.sizes}")
```
This change is a bit silly, but flake will fail if it detects something like `not found in` even if it is just text

k-dominik commented on tiktorch/server/session/process.py:
```diff
@@ -91,13 +96,21 @@ def create_dataset(self, mean, stddev):
+            raise ValueError("Server isn't initialized")
+        return self._worker
+
+        return Shutdown()
```
unreachable code.

k-dominik commented on tiktorch/server/session/process.py:
```diff
@@ -110,24 +123,27 @@ def _run_model_session_process(
+) -> Tuple[_mp.Process, BioModelClient]:
     model_descr = _get_model_descr_from_model_bytes(model_bytes)
+    proc, api = start_process(interface_class=ModelSessionProcess, log_queue=log_queue)
     return proc, BioModelClient(
         input_specs=model_descr.inputs,
         output_specs=model_descr.outputs,
```
would be cool to fix the type hinting here.

k-dominik reviewed (approved):
Looks fine, also very sensible changeset.

thodkatz commented:
In my last commit I have introduced one more suggestion to close session if initialization fails, so the client can request a new session without the devices being reserved. I have added one test to illustrate this as well.
